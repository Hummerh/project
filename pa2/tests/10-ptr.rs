fn main() {
   let mut x : i32 = 5;
   let mut b : bool = true;

   let y : &i32 = &7;
   let c : &bool = &false;

   let z : & &i32 = &y;
   let d : & &bool = &c;

   x = **z + **z;
   b = **d || (**z > 3);

   let p : Box<&mut i32> = Box::new(&mut x);
   let q : i32 = **p;
}
