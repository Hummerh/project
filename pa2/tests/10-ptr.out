(crate
    (items
        (fn-def
            (id
                (main))
            (block
                (let
                    (pat-mut-id
                        (id
                            (x)))
                    (type-i32)
                    (lit-dec))
                (let
                    (pat-mut-id
                        (id
                            (b)))
                    (type-bool)
                    (true))
                (let
                    (pat-id
                        (id
                            (y)))
                    (type-ref
                        (type-i32))
                    (addr-of
                        (lit-dec)))
                (let
                    (pat-id
                        (id
                            (c)))
                    (type-ref
                        (type-bool))
                    (addr-of
                        (false)))
                (let
                    (pat-id
                        (id
                            (z)))
                    (type-ref
                        (type-ref
                            (type-i32)))
                    (addr-of
                        (id
                            (y))))
                (let
                    (pat-id
                        (id
                            (d)))
                    (type-ref
                        (type-ref
                            (type-bool)))
                    (addr-of
                        (id
                            (c))))
                (assign
                    (id
                        (x))
                    (add
                        (deref
                            (deref
                                (id
                                    (z))))
                        (deref
                            (deref
                                (id
                                    (z))))))
                (assign
                    (id
                        (b))
                    (or
                        (deref
                            (deref
                                (id
                                    (d))))
                        (gt
                            (deref
                                (deref
                                    (id
                                        (z))))
                            (lit-dec))))
                (let
                    (pat-id
                        (id
                            (p)))
                    (type-box
                        (type-ref-mut
                            (type-i32)))
                    (box-new
                        (exprs
                            (addr-of-mut
                                (id
                                    (x))))))
                (let
                    (pat-id
                        (id
                            (q)))
                    (type-i32)
                    (deref
                        (deref
                            (id
                                (p)))))))))